cli/cook/subcommands/tail.py

```diff
@@ -0,0 +1,159 @@
+import logging
+import os
+import time
+from functools import partial
+
+from cook import mesos, http
+from cook.querying import query_unique_and_run
+from cook.util import strip_all, check_positive
+
+CHUNK_SIZE = 4096
+LINE_DELIMITER = '\n'
+DEFAULT_NUM_LINES = 10
+DEFAULT_FOLLOW_SLEEP_SECS = 1.0
+
+
+# For everything we print in tail, we want to forcibly flush
+# the stream, and we don't want to end with a newline
+prn = partial(print, flush=True, end='')
+
+
+def read_file(instance, sandbox_dir, path, offset=None, length=None):
+    """Calls the Mesos agent files/read API for the given path, offset, and length"""
+    logging.info(f'reading file from sandbox {sandbox_dir} with path {path} at offset {offset} and length {length}')
+    agent_url = mesos.instance_to_agent_url(instance)
+    params = {'path': os.path.join(sandbox_dir, path or '')}
+    if offset is not None:
+        params['offset'] = offset
+    if length is not None:
+        params['length'] = length
+
+    resp = http.__get(f'{agent_url}/files/read', params=params)
+    if resp.status_code == 404:
+        raise Exception(f"Cannot open '{path}' for reading (file was not found).")
+
+    if resp.status_code != 200:
+        logging.error(f'mesos agent returned status code {resp.status_code} and body {resp.text}')
+        raise Exception('Encountered error when reading file from Mesos agent.')
+
+    return resp.json()
+
+
+def print_lines(lines):
+    """Prints the given list of lines, delimited, and with no trailing newline"""
+    prn(LINE_DELIMITER.join(lines))
+
+
+def tail_for_instance(instance, sandbox_dir, path, num_lines_to_print, follow, follow_sleep_seconds):
+    """
+    Tails the contents of the Mesos sandbox path for the given instance. The
+    algorithm reads chunks backwards from the end of the file and splits them into
+    lines as it goes. If it finds that enough lines have been read to satisfy the
+    user's request, or if it reaches the beginning of the file, it stops (unless
+    follow is truthy). If follow is truthy, it will try and read more data from the
+    file until the user terminates. Note that this assumes files will not shrink.
+    """
+    read = partial(read_file, instance=instance, sandbox_dir=sandbox_dir, path=path)
+
+    # Get the current file size
+    file_size = read()['offset']
+
+    # Initialize loop variables
+    offset = max(file_size - CHUNK_SIZE, 0)
+    length = file_size - offset
+    text_buffer = ''
```

Comment:
Renaming `text_buffer` to something like `partial_line_buffer` would have helped me understand this code more easily.

Reply:
Done